hub/core/chunk.py

```diff
@@ -0,0 +1,141 @@
+                `ShapeEncoder` handles encoding the `start_byte` and `end_byte` for each sample.
+                `BytePositionsEncoder` handles encoding the `shape` for each sample.
+
+            To see how this chunk is composed into bytes, check out `tobytes`.
+            To see how this chunk is constructed from bytes, checkout `frombuffer`.
+
+        Args:
+            encoded_shapes (np.ndarray): Used to construct `ShapeEncoder` if this chunk already exists. Defaults to None.
+            encoded_byte_positions (np.ndarray): Used to construct `BytePositionsEncoder` if this chunk already exists. Defaults to None.
+            data (memoryview, optional): If this chunk already exists, data should be set. Defaults to None.
+        """
+
+        # data
+        self._data: Union[bytearray, memoryview] = data or bytearray()
+
+        # header
+        self.shapes_encoder = ShapeEncoder(encoded_shapes)
+        self.byte_positions_encoder = BytePositionsEncoder(encoded_byte_positions)
+
+    @property
+    def memoryview_data(self):
+        return memoryview(self._data)
+
+    @property
+    def num_data_bytes(self):
+        return len(self._data)
+
+    def is_under_min_space(self, min_data_bytes_target: int):
+        return self.num_data_bytes < min_data_bytes_target
+
+    def has_space_for(self, num_bytes: int, max_data_bytes: int):
+        return self.num_data_bytes + num_bytes <= max_data_bytes
+
+    def append_sample(
+        self, incoming_buffer: memoryview, max_data_bytes: int
+    ) -> Tuple["Chunk"]:
+        """Store `incoming_buffer` in this chunk.
+
+        Args:
+            incoming_buffer (memoryview): Buffer that represents a single sample.
+            max_data_bytes (int): Used to determine if this chunk has space for `incoming_buffer`.
+
+        Raises:
+            FullChunkError: If `incoming_buffer` is too large.
+        """
+
+        incoming_num_bytes = len(incoming_buffer)
+
+        if not self.has_space_for(incoming_num_bytes, max_data_bytes):
+            raise FullChunkError(
+                f"Chunk does not have space for the incoming bytes (incoming={incoming_num_bytes}, max={max_data_bytes})."
+            )
+
+        # note: incoming_num_bytes can be 0 (empty sample)
+        self._data += incoming_buffer
+
+    def update_headers(
+        self, incoming_num_bytes: int, num_samples: int, sample_shape: Sequence[int]
+    ):
+        """Updates this chunk's header. A chunk should NOT exist without headers.
+
+        Args:
+            incoming_num_bytes (int): Number of bytes this header should account for. Should be divisble by `num_samples`.
+            num_samples (int): Number of samples this header should account for.
+            sample_shape (Sequence[int]): Every sample that `num_samples` symbolizes is considered to have `sample_shape`.
+
+        Raises:
+            Exception: If trying to update headers when no data was actually added.
```

Comment:
Does this still happen? If so, we should throw something more meaningful than base `Exception`